src/wanna/core/services/notebook.py

```diff
@@ -393,3 +408,84 @@ def build(self) -> int:
+        """
+        Figuring out the diff between GCP and wanna.yaml. Lists user-managed notebooks to be deleted and created.
+        """
+        # We list compute instances and not notebooks, because with notebooks you cannot list instances in all zones.
+        # So instead we list all Compute Engine instances with notebook labels
+        instance_client = compute_v1.InstancesClient()
+        request = compute_v1.AggregatedListInstancesRequest(
+            filter=f"(labels.wanna_resource:notebook) (labels.wanna_project:{self.wanna_project.name})"
+        )
+        request.project = self.config.gcp_profile.project_id
+        agg_list = instance_client.aggregated_list(request=request)
+        gcp_instances = itertools.chain(*[resp.instances for zone, resp in agg_list if resp.instances])
+
+        active_notebooks = [{"name": i.name, "zone": i.zone.split("/")[-1]} for i in gcp_instances]
+        wanna_notebooks = [{"name": i.name, "zone": i.zone, "model": i} for i in self.instances]
+        active_notebook_names = [n.get("name") for n in active_notebooks]
+        wanna_notebook_names = [n.get("name") for n in wanna_notebooks]
+
+        to_be_deleted = []
+        to_be_created = []
+        """
+        Notebooks to be deleted
+        """
+        for active_notebook in active_notebooks:
+            if active_notebook.get("name") not in wanna_notebook_names:
+                to_be_deleted.append(active_notebook)
+        """
+        Notebooks to be created
+        """
+        for wanna_notebook in self.instances:
+            if wanna_notebook.name not in active_notebook_names:
+                to_be_created.append(wanna_notebook)
+        return to_be_deleted, to_be_created
+
+    def sync(self, force: bool, skip_containers: bool = False) -> None:
+        """
+        1. Reads current notebooks where label is defined per field wanna_project.name in wanna.yaml
+        2. Does a diff between what is on GCP and what is on yaml
+        3. Delete the ones in GCP that are not in wanna.yaml
+        4. Create the ones defined in yaml and missing in GCP
+        """
+        to_be_deleted, to_be_created = self._return_diff()
+
+        if to_be_deleted:
+            to_be_deleted_str = "\n".join(["- " + item["name"] for item in to_be_deleted])
+            logger.user_info(f"Notebooks to be deleted:\n{to_be_deleted_str}")
+            should_delete = True if force else typer.confirm("Are you sure you want to delete them?")
+            if should_delete:
+                for notebook in to_be_deleted:
+                    with logger.user_spinner(f"Deleting {notebook.get('name')}"):
+                        self.notebook_client.delete_instance(
+                            name=f"projects/{self.config.gcp_profile.project_id}/locations/"
+                            f"{notebook.get('zone')}/instances/{notebook.get('name')}"
+                        ).result()
+
+        if to_be_created:
```

Comment:
I am not 100% sure but isn't this same logic elsewhere ? can it be re-used ?

Reply:
Yes, I am doing a small refactor and then it can be reused :)